pilot/pkg/xds/grpc.go

```diff
@@ -0,0 +1,83 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"context"
+
+	middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/keepalive"
+
+	"istio.io/istio/pilot/pkg/features"
+	istiokeepalive "istio.io/istio/pkg/keepalive"
+)
+
+type (
+	SendHandler  func(errorChan chan error)
+	ErrorHandler func(err error)
+)
+
+var timeout = features.XdsPushSendTimeout
+
+// Send with timeout if specified. If timeout is zero, sends without timeout.
+func Send(ctx context.Context, send SendHandler, errorh ErrorHandler) error {
+	errChan := make(chan error, 1)
```

Comment:
I would rewrite this as

```go
go func() {
  err := send()
  errChan <- err
  close(errChan)
}
```

Then the calls don't have to worry about channels. They all just send on the channel as the last action anyways?